lib/ClangImporter/ImportType.cpp

```diff
@@ -2252,114 +2403,37 @@ ParameterList *ClangImporter::Implementation::importFunctionParameterList(
-                                        ? PTK_UnsafePointer
-                                        : PTK_UnsafeMutablePointer;
-      auto genericType =
-          findGenericTypeInGenericDecls(*this, templateParamType, genericParams,
-                                        attrs, paramAddDiag);
-      swiftParamTy = genericType->wrapInPointer(pointerKind);
-      if (!swiftParamTy)
-        return nullptr;
-    } else if (isa<clang::ReferenceType>(paramTy) &&
-               isa<clang::TemplateTypeParmType>(paramTy->getPointeeType())) {
-      auto templateParamType =
-          cast<clang::TemplateTypeParmType>(paramTy->getPointeeType());
-      swiftParamTy =
-          findGenericTypeInGenericDecls(*this, templateParamType, genericParams,
-                                        attrs, paramAddDiag);
-      if (!paramTy->getPointeeType().isConstQualified())
-        isInOut = true;
-    } else if (auto *templateParamType =
-                   dyn_cast<clang::TemplateTypeParmType>(paramTy)) {
-      swiftParamTy =
-          findGenericTypeInGenericDecls(*this, templateParamType, genericParams,
-                                        attrs, paramAddDiag);
-    } else if (auto refType = dyn_cast<clang::ReferenceType>(paramTy)) {
-      // We don't support reference type to a dependent type, just bail.
-      if (refType->getPointeeType()->isDependentType()) {
-        addImportDiagnostic(
-            param, Diagnostic(diag::parameter_type_not_imported, param),
-            param->getSourceRange().getBegin());
-        return nullptr;
-      }
 
-      paramTy = refType->getPointeeType();
-      if (!paramTy.isConstQualified())
-        isInOut = true;
-    }
+    bool isInOut = false;
+    bool isParamTypeImplicitlyUnwrapped = false;
 
+    auto swiftParamTy = importParameterType(
+        param, optionalityOfParam, importKind, allowNSUIntegerAsInt,
+        /*isNSDictionarySubscriptGetter=*/false,
+        /*paramIsError=*/false,
+        /*paramIsCompletionHandler=*/false,
+        /*completionHandlerErrorParamIndex=*/None, genericParams, paramAddDiag,
+        isInOut, isParamTypeImplicitlyUnwrapped);
     if (!swiftParamTy) {
-      auto importedType = importType(paramTy, importKind, paramAddDiag,
-                                     allowNSUIntegerAsInt, Bridgeability::Full,
-                                     attrs, OptionalityOfParam);
-      if (!importedType) {
-        addImportDiagnostic(
-            param, Diagnostic(diag::parameter_type_not_imported, param),
-            param->getSourceRange().getBegin());
-        return nullptr;
-      }
-
-      isParamTypeImplicitlyUnwrapped = importedType.isImplicitlyUnwrapped();
-      swiftParamTy = importedType.getType();
+      addImportDiagnostic(param,
+                          Diagnostic(diag::parameter_type_not_imported, param),
+                          param->getSourceRange().getBegin());
+      return nullptr;
     }
 
```

Comment:
Can we add something like `auto swiftParamTy = *swiftParamTyOpt;` here so that the lines extracting the optional can be left alone?